paillier/paillier.go

```diff
@@ -0,0 +1,234 @@
+		g:       pub.GetG(),
+	}
+}
+
+func (pub *PublicKey) Encrypt(m *big.Int) (*big.Int, error) {
+	// Ensure 0 <= m < n
+	if m.Sign() < 0 {
+		return nil, ErrInvalidMessage
+	}
+	if m.Cmp(pub.n) >= 0 {
+		return nil, ErrInvalidMessage
+	}
+
+	// gcd(r, n)=1
+	r, err := RandomCoprimeInt(pub.n)
+	if err != nil {
+		return nil, err
+	}
+
+	// c = (g^m * r^n) mod n^2
+	gm := new(big.Int).Exp(pub.g, m, pub.nSquare) // g^m
+	rn := new(big.Int).Exp(r, pub.n, pub.nSquare) // r^n
+	c := new(big.Int).Mul(gm, rn)
+	c = c.Mod(c, pub.nSquare)
+	return c, nil
+}
+
+// https://en.wikipedia.org/wiki/Paillier_cryptosystem
+// Add implements homomorphic addition of plaintexts
+func (pub *PublicKey) Add(encA *big.Int, encB *big.Int) *big.Int {
+	encAB := new(big.Int).Mul(encA, encB)
+	ret := new(big.Int).Mod(encAB, pub.nSquare)
+	return ret
+}
+
+// https://en.wikipedia.org/wiki/Paillier_cryptosystem
+// Mul implements homomorphic multiplication of plaintexts
+func (pub *PublicKey) Mul(encA *big.Int, scalar *big.Int) *big.Int {
+	return new(big.Int).Exp(encA, scalar, pub.nSquare)
+}
+
+// https://en.wikipedia.org/wiki/Paillier_cryptosystem
+// Paillier implements the Paillier crypto system
+type Paillier struct {
+	// public key
+	*PublicKey
+
+	// private key
+	lambda *big.Int // λ=lcm(p−1, q−1)
+	mu     *big.Int // μ=(L(g^λ mod n^2))^-1 mod n
+}
+
+func NewPaillier(keySize int) (*Paillier, error) {
+	n, lambda, err := getNAndLambda(keySize)
+	if err != nil {
+		return nil, err
+	}
+	nSquare := new(big.Int).Mul(n, n) // n^2
+	g, mu, err := getGAndMu(lambda, n, nSquare)
+	if err != nil {
+		return nil, err
+	}
+	return &Paillier{
+		lambda: lambda,
+		mu:     mu,
+		PublicKey: &PublicKey{
+			n:       n,
+			nSquare: nSquare,
+			g:       g,
+		},
+	}, nil
+}
+
+// Decrypt computes the plaintext from the ciphertext
+func (p *Paillier) Decrypt(c *big.Int) (*big.Int, error) {
+	// Ensure 0 < c < n^2
+	if c.Sign() <= 0 {
+		return nil, ErrInvalidMessage
+	}
+	if c.Cmp(p.PublicKey.nSquare) >= 0 {
+		return nil, ErrInvalidMessage
+	}
+
+	// c and n should be relative prime
+	if !IsRelativePrime(c, p.PublicKey.n) {
+		return nil, ErrInvalidMessage
+	}
+
+	x := new(big.Int).Exp(c, p.lambda, p.PublicKey.nSquare)
+	l, err := lFunction(x, p.PublicKey.n)
+	if err != nil {
+		return nil, err
+	}
+	l = l.Mul(l, p.mu)
+	l = l.Mod(l, p.PublicKey.n)
+	return l, nil
+}
+
+func (p *Paillier) Copy() *Paillier {
+	return &Paillier{
+		lambda:    new(big.Int).Set(p.lambda),
+		mu:        new(big.Int).Set(p.mu),
+		PublicKey: p.PublicKey.Copy(),
+	}
+}
+
+// getNAndLambda returns N and lambda.
+// n = pq and lambda = lcm(p-1, q-1)
+func getNAndLambda(keySize int) (*big.Int, *big.Int, error) {
+	pqSize := keySize / 2
+	for i := 0; i < maxGenN; i++ {
+		// random two primes p and q
+		p, err := rand.Prime(rand.Reader, pqSize)
+		if err != nil {
+			return nil, nil, err
+		}
+		q, err := rand.Prime(rand.Reader, pqSize)
+		if err != nil {
+			return nil, nil, err
+		}
+
```

Comment:
Due to correct, maybe need to check gcd(p,q)=1. However, in this case, we consider that p and q are primes with equal lengths, so only need to check "p !=q". 
In fact, the probability of p=q happened is very low about 2^(-pqSize). 
In our application, this value is 2^(-1024).



Well, this is my mistake.

Reply:
OK, so we just meed to check if p == q.

Reply:
Yes !